# QUIZ 1 RO-2 (Q)
by: Sumihar Christian 

secara default scipy.minize menggunakan method ```Sequential Least SQuares Programming```, which basically sebuah method iterative buat menyelesaikan constrained linear optimization.  

Algoritmanya mirip sama punya KKT (they solve the same stuff.): 

$$
\min_x f(x) \\ 
constraint: \\
b(x) >= 0 \\
c(x) = 0 \\ \\ 
\mathcal{L}(x,\lambda,\sigma) = f(x) - \lambda^{T}b(x)-\sigma^{T}c(x)
$$

$$ f(x):(x_1-1)^2 + ((x_2+2)^2 + (x_3-2)^2)
$$ 

constraint to: 
$$ g(x): (2x_1 + 3x_2 -1) = 0 \\
h(x): (x_1 + x_2 + 2x_3 - 4) = 0 
$$

initial guess ($x_0$) : 
$$ x_0 = (5,-3,1) $$
that suffices the constraint

In [1]:
import numpy as np 
from scipy.optimize import minimize

In [6]:
def cons1(x): 
    return 2*x[0] + 3*x[1] -1
def cons2(x): 
    return x[0] + x[1] + 2*x[2] - 4 
def obj(x): 
    return ((x[0]-1)**2 + (x[1]+2)**2 + (x[2]-2)**2)

In [7]:
con1 = {
    'type':'eq', 
    'fun':cons1
}
con2 = {
    'type':'eq',
    'fun':cons2
}
constraint = ([con1,con2])

In [8]:
x0 =(5,-3,1) #initial guess
sol = minimize(obj,x0,constraints=constraint)

In [9]:
sol

     fun: 2.1320754716981147
     jac: array([ 1.43396229,  2.3773585 , -0.90566036])
 message: 'Optimization terminated successfully.'
    nfev: 11
     nit: 2
    njev: 2
  status: 0
 success: True
       x: array([ 1.71698113, -0.81132075,  1.54716981])

# results
objective value = 2.1320754716981147   
$x_1$ = 1.17169    
$x_2$ = -0.811132075    
$x_3$ = 1.54716981    

![](https://pics.me.me/when-your-friends-tell-you-about-how-awesome-it-is-29157590.png)

penjelasan scipy.minimize bisa dilihat [disini](https://github.com/svmihar/ro2/blob/master/quiz_bayangan/jawaban_kuis.ipynb) 